Remove suggestion about iteration count in coerce

gurry opened the pull request:
Fixes #122561 

The iteration count-centric suggestion was implemented in PR #100094, but it was based on the wrong assumption that the type mismatch error depends on the number of times the loop iterates. As it turns out, that is not true (see this comment for details: https://github.com/rust-lang/rust/pull/122679#issuecomment-2017432531)

This PR attempts to remedy the situation by changing the suggestion from the one centered on iteration count to a simple suggestion to add a return value.

It should also fix #100285 by simply making it redundant.

bors commented:
:umbrella: The latest upstream changes (presumably #123006) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

gurry commented:
> ☔ The latest upstream changes (presumably #123006) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Rebased

estebank commented on tests/ui/typeck/issue-98982.stderr:
```diff
@@ -8,26 +8,11 @@ LL | |         return i;
    | |_____^ expected `i32`, found `()`
    |
-note: the function expects a value to always be returned, but loops might run zero times
-  --> $DIR/issue-98982.rs:2:5
-   |
-LL |     for i in 0..0 {
-   |     ^^^^^^^^^^^^^ this might have zero elements to iterate on
-LL |         return i;
-   |         -------- if the loop doesn't execute, this value would never get returned
-help: return a value for the case when the loop has zero elements to iterate on
+help: consider returning a value here
    |
 LL ~     }
 LL ~     /* `i32` value */
    |
-help: otherwise consider changing the return type to account for that possibility
-   |
-LL ~ fn foo() -> Option<i32> {
-LL |     for i in 0..0 {
-LL ~         return Some(i);
-LL ~     }
-LL ~     None
-   |
```
I don't think that this reversal is reasonable, because the explanation *for this case* is exactly right: `for i in 0..0` will never iterate, right?

gurry replied:
Yes, `for i in 0..0` will never iterate, but that is not what is causing the error. `for i in 0..100` will definitely iterate but the same error still appears.

estebank commented on tests/ui/typeck/issue-100285.stderr:
```diff
@@ -12,47 +12,11 @@ LL | |
 LL | |     }
    | |_____^ expected `i32`, found `()`
    |
-note: the function expects a value to always be returned, but loops might run zero times
-  --> $DIR/issue-100285.rs:2:5
-   |
-LL |     for i in 0..0 {
-   |     ^^^^^^^^^^^^^ this might have zero elements to iterate on
```
Same here. We should even be detecting explicitly known to never iterate ranges, where the start and the end of a `..` range are the same, because the entire thing is *for sure* dead code.

gurry replied:
Yes, it would be a good idea to warn the user that `for i in 0..0 { ... }` is dead code. Maybe I can do that in a different PR.

estebank commented on compiler/rustc_hir_typeck/src/coercion.rs:
```diff
@@ -1819,102 +1788,61 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
+        };
+
+        if let Some(span) = span {
+            let (msg, suggestion) = if ty.is_never() {
+                (
+                    "consider adding a diverging expression here",
+                    "`loop {}` or `panic!(\"...\")`".to_string(),
+                )
```
This is a great additional bit of context.

estebank reviewed (commented):
I'm not entirely on board with completely removing the previous message, because I feel that we'd be giving an obscure message still. Instead I would *add* logic to it in order to detect a few cases:

1. when the range is known to never run, mention that and modify the other messages to account for it so they make sense
2. when there's `loop { return val; }` with no conditions, to explain that "rustc is not smart enough to understand that `val` will always be returned" (although ideally we'd have a lang semantic change to account for that so that we don't error at all in that case)
3. when there's a `loop { if condition { return val; } }` with the current explanation (tweaked to make it clearer that it's not just the loop might not run even once, but also that it might run multiple times but never reach the return before the loop breaks)

gurry commented:
> I'm not entirely on board with completely removing the previous message, because I feel that we'd be giving an obscure message still. Instead I would _add_ logic to it in order to detect a few cases:

I agree that we should provide more context. However, any notes and explanations that talk about the loop iteration count are flawed IMHO because the error has _nothing_ to do with loop iteration count.

For example, below snippet in which the loop runs zero times triggers the type mismatch error:
```rust
fn test() -> bool {
    for i in 0..0 {
        return true;
    }
}
```
but so does this one in which the loop runs 5 times:
```rust
fn test() -> bool {
    for i in 0..5 {
        return true;
    }
}
```
(Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ff4e48a30ffbc4f0f344bd6ccae27d5f)

The reality is that the error actually stems from the peculiar structure of the `for` loop desugaring not the whether the loop iterates or not (see this comment for details: https://github.com/rust-lang/rust/pull/122679#issuecomment-2017432531)

In light of this underlying reality, the current set of notes (like "the function expects a value to always be returned, but loops might run zero times") are incorrect and somewhat misleading because they seem to attribute the error to the loop iteration count.

If we must add any explanation, it should be about the for loop desugaring which is the actual cause. However, I couldn't find any good way to word it which would not get into the compiler internals. If you can suggestion a user friendly text I'd love to add that.

> 
> 1. when the range is known to never run, mention that and modify the other messages to account for it so they make sense

I actually had done just that in my now abandoned PR #122679, wherein I computed the actual iteration count for the loop and showed different notes depending on that, but then I realised no amount sophistication in suggestions centered on loop iteration count actually help here because the cause of the error is entirely something else.

> 2. when there's `loop { return val; }` with no conditions, to explain that "rustc is not smart enough to understand that `val` will always be returned" (although ideally we'd have a lang semantic change to account for that so that we don't error at all in that case)
> 3. when there's a `loop { if condition { return val; } }` with the current explanation (tweaked to make it clearer that it's not just the loop might not run even once, but also that it might run multiple times but never reach the return before the loop breaks)

Oddly enough, the plain vanilla `loop` actually doesn't trigger the type mismatch error. Both cases 2. and 3. you mentioned compiles without any error as you can see in this Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c2bc903c88d86cd88cb23e9b88f99d88). So there's no question of adding any explanation for them.

estebank commented:
Thanks! The comment on the other PR gives appropriate context. Now I'm trying to think what the appropriate explanation of the situation might be. It honestly might be reasonable to explain "yes, this desugars to something else that evaluates to `()`, sorry".

gurry commented:
> It honestly might be reasonable to explain "yes, this desugars to something else that evaluates to `()`, sorry".

"Desugar" may not be a term well understood outside of compiler engineering. Maybe something along the lines of, "for loops always evaluate to type `()`".

estebank commented on tests/ui/did_you_mean/compatible-variants.stderr:
```diff
@@ -19,9 +20,14 @@ LL +     None
+   |
+LL ~     }
+LL +     /* `Option<()>` value */
+   |
```
Here we are providing effectively the same suggestion twice. We need to find a way to avoid one or the other (or ideally, unify the logic to a single place, particularly because the pre-existing suggestion gives you the actual value to use in this case).

gurry replied:
Done.

The suggestion is now invoked from `emit_type_mismatch_suggestions()` in `demand.rs` which ensures it will not be called if the other suggestion has already been emitted.

gurry replied:
@estebank please review

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8615604583/job/23611493430) [(plain)](https://github.com/rust-lang/rust/commit/2b211b1ffbcd7863c0eefcbdfb5a4509fc1572a0/checks/23611493430/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.4s done
#16 DONE 33.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
     |          ^^^^^^
     |
    ::: compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs:41:1
     |
41   | pub(crate) struct FnCtxt<'a, 'tcx> {
     | ---------------------------------- similarly named struct `FnCtxt` defined here
help: a struct with a similar name exists
     |
     |
1306 |     tcx: FnCtxt<'tcx>,
help: consider importing one of these items
     |
38   + use crate::TyCtxt;
     |
```

</details>

estebank commented:
@bors r+

bors commented:
:pushpin: Commit fa73e370e59f18d5eaab6021aff112446e81a727 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank fa73e370e59f18d5eaab6021aff112446e81a727 -->
<!-- homu: {"type":"Approved","sha":"fa73e370e59f18d5eaab6021aff112446e81a727","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit fa73e370e59f18d5eaab6021aff112446e81a727 with merge b16d83c110a65257c4209ded155c61671da6a08b...
<!-- homu: {"type":"BuildStarted","head_sha":"fa73e370e59f18d5eaab6021aff112446e81a727","merge_sha":"b16d83c110a65257c4209ded155c61671da6a08b"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8753173541/job/24022673865)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8753173541/job/24022673865","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`armhf-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8753173541/job/24022343629) [(plain)](https://github.com/rust-lang-ci/rust/commit/b16d83c110a65257c4209ded155c61671da6a08b/checks/24022343629/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

gurry committed:
Remove note about iteration count in coerce

and replace it with a simple note suggesting
returning a value.

The type mismatch error was never due to
how many times the loop iterates. It is more
because of the peculiar structure of what the for
loop desugars to. So the note talking about
iteration count didn't make sense

gurry commented:
Rebased to fix build failure.

bors commented:
:umbrella: The latest upstream changes (presumably #124398) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

gurry commented:
> ☔ The latest upstream changes (presumably #124398) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Rebased

gurry commented:
@estebank Failing build fixed. Please review again

Dylan-DPC commented:
@bors r=estebank

bors commented:
:pushpin: Commit 6289ed842897f75f7b1f06aee44f9f123f57077d has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 6289ed842897f75f7b1f06aee44f9f123f57077d -->
<!-- homu: {"type":"Approved","sha":"6289ed842897f75f7b1f06aee44f9f123f57077d","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/124730#issuecomment-2094643372) for pull requests below priority 9999. This pull request will be tested once the tree is reopened.

bors commented:
:hourglass: Testing commit 6289ed842897f75f7b1f06aee44f9f123f57077d with merge 06e88c306a180067b6798901c26828d8ed7b0963...
<!-- homu: {"type":"BuildStarted","head_sha":"6289ed842897f75f7b1f06aee44f9f123f57077d","merge_sha":"06e88c306a180067b6798901c26828d8ed7b0963"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8958378760/job/24604145377)
Approved by: estebank
Pushing 06e88c306a180067b6798901c26828d8ed7b0963 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"estebank","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8958378760/job/24604145377"},"merge_sha":"06e88c306a180067b6798901c26828d8ed7b0963"} -->

rust-timer commented:
Finished benchmarking commit (06e88c306a180067b6798901c26828d8ed7b0963): [comparison URL](https://perf.rust-lang.org/compare.html?start=02f7806ecd641d67c8f046b073323c7e176ee6d2&end=06e88c306a180067b6798901c26828d8ed7b0963&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=02f7806ecd641d67c8f046b073323c7e176ee6d2&end=06e88c306a180067b6798901c26828d8ed7b0963&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=02f7806ecd641d67c8f046b073323c7e176ee6d2&end=06e88c306a180067b6798901c26828d8ed7b0963&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.8%  | [3.8%, 3.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -5.2% | [-5.2%, -5.2%] | 1     |
| All ❌✅ (primary)                 | 1.5%  | [-0.7%, 3.8%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=02f7806ecd641d67c8f046b073323c7e176ee6d2&end=06e88c306a180067b6798901c26828d8ed7b0963&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=02f7806ecd641d67c8f046b073323c7e176ee6d2&end=06e88c306a180067b6798901c26828d8ed7b0963&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 676.151s -> 673.364s (-0.41%)
**Artifact size**: 315.84 MiB -> 315.81 MiB (-0.01%)
